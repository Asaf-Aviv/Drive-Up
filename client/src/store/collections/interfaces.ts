import { MovieShowcase } from '../movies/interfaces';

export interface Collection {
  id: number;
  name: string;
  overview: string;
  poster_path: string | null;
  backdrop_path: string | null;
  parts: MovieShowcase[];
}
