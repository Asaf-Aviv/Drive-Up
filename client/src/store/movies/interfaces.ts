export interface Results<T> {
  results: T;
  page: number;
  total_pages: number;
  total_results: number;
}

export interface Genre {
  id: number;
  name: string;
}

export interface Series {
  id: number;
  name: string;
  poster_path: string;
  backdrop_path: string;
}

export interface Company {
  name: string;
  id: number;
  logo_path: string | null;
  origin_country: string;
}

export interface Image {
  file_path: string;
  width: number;
  height: number;
}

export interface Video {
  id: string;
  key: string;
  name: string;
  site: string;
  type: string;
}

export interface MovieShowcase {
  poster_path: string | null;
  adult: boolean;
  overview: string;
  release_date: string;
  genre_ids: number[];
  id: number;
  original_title: string;
  original_language: string;
  title: string;
  backdrop_path: string | null;
  popularity: number;
  vote_count: number;
  video: boolean;
  vote_average: number;
}

export interface Movie {
  adult: boolean;
  backdrop_path: string | null;
  belongs_to_collection: Series | null;
  budget: number;
  genres: Genre[];
  homepage: string | null;
  id: number;
  images: {
    posters: Image[];
    backdrops: Image[];
  };
  videos: {
    results: Video[];
  };
  similar: Results<MovieShowcase[]>;
  recommendations: Results<MovieShowcase[]>;
  original_language: string;
  original_title: string;
  overview: string;
  popularity: number;
  poster_path: string | null;
  production_companies: Company[];
  release_date: string;
  revenue: number;
  runtime: number | null;
  tagline: string | null;
  title: string;
  video: boolean;
  vote_average: number;
  vote_count: number;
}
