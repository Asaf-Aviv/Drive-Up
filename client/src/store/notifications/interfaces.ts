type NotificationVariant = 'error' | 'success' | 'info' | 'warning';

export interface Notification {
  id: string;
  variant: NotificationVariant;
  message: string;
}
