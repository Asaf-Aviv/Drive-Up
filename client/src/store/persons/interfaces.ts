export interface PersonSummary {
  profile_path: string;
  adult: boolean;
  id: number;
  name: string;
  popularity: number;
  known_for_department: string;
  gender: 0 | 1 | 2;
}
