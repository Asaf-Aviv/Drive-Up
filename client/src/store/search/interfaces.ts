import { MovieShowcase } from '../movies/interfaces';
import { PersonSummary } from '../persons/interfaces';
import { ShowShowcase } from '../shows/interfaces';

interface MediaType<T = 'movie' | 'tv' | 'person'> {
  media_type: T;
}

export type SearchResults =
  | MovieShowcase & MediaType<'movie'>
  | ShowShowcase & MediaType<'tv'>
  | PersonSummary & MediaType<'person'>;
