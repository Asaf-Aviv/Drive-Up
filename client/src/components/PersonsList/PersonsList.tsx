import React, { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { List, ListItem } from '@material-ui/core';
import useShallowEqualSelector from '../../hooks/useShallowEqualSelector';
import { useOnVisibilityTrigger } from '../../hooks/useOnVisibilityTrigger';
import { PersonSummary as IPersonSummary } from '../../store/persons/interfaces';
import RetryButton from '../RetryButton';
import { requestPersons } from '../../store/persons/actions';
import PersonShowcase from '../PersonShowcase';

const PersonsList: React.FC = () => {
  const {
    page, loading, persons, error,
  } = useShallowEqualSelector(state => state.persons);
  const dispatch = useDispatch();
  const fetchNextPageTrigger = useRef<HTMLElement>(null);

  const fetchNextPersonsPage = () => {
    dispatch(requestPersons(page + 1));
  };

  useEffect(fetchNextPersonsPage, []);

  useOnVisibilityTrigger(fetchNextPageTrigger, fetchNextPersonsPage);

  const renderPersonSummary = (person: IPersonSummary) => (
    <ListItem key={person.id}>
      <PersonShowcase {...person} />
    </ListItem>
  );

  return (
    <main>
      <List>
        {persons.map(renderPersonSummary)}
        {!error && !loading && <span ref={fetchNextPageTrigger} />}
        {loading && <span>Loading...</span>}
        {error && <RetryButton onClick={fetchNextPersonsPage} />}
      </List>
    </main>
  );
};

export default PersonsList;
