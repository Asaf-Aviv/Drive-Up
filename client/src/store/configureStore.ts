import {
  createStore,
  applyMiddleware,
  Middleware,
  Reducer,
} from 'redux';
import thunkMiddleware from 'redux-thunk';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import rootSaga from './sagas/rootSaga';

export default function configureStore(rootReducer: Reducer, preloadedState = {}) {
  const devMode = process.env.NODE_ENV !== 'production';

  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware, thunkMiddleware];

  if (devMode) {
    middlewares.push(logger);
  }

  const middleWareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(
    rootReducer,
    preloadedState,
    devMode
      ? composeWithDevTools(middleWareEnhancer)
      : middleWareEnhancer,
  );

  sagaMiddleware.run(rootSaga);

  return store;
}
