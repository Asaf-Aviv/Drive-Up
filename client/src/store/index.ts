import {
  combineReducers,
  createStore,
  applyMiddleware,
  Middleware,
} from 'redux';
import thunkMiddleware from 'redux-thunk';
import logger from 'redux-logger';
import createSagaMiddleware, { SagaMiddleware } from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import notificationsReducer from './notifications/reducers';
import moviesReducer from './movies/reducers';

const rootReducer = combineReducers({
  notifications: notificationsReducer,
  movies: moviesReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default () => {
  const devMode = process.env.NODE_ENV !== 'production';

  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware, thunkMiddleware];

  if (devMode) {
    middlewares.push(logger);
  }

  const middleWareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(
    rootReducer,
    devMode
      ? composeWithDevTools(middleWareEnhancer)
      : middleWareEnhancer,
  );

  // sagaMiddleware.run()

  return store;
};
