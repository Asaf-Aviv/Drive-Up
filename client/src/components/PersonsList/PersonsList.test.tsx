import React from 'react';
import 'intersection-observer';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import { MemoryRouter } from 'react-router';
import { MovieShowcase } from '../../store/movies/interfaces';
import configureStore from '../../store/configureStore';
import moviesReducer from '../../store/movies/reducers';
import { rootReducer } from '../../store';
import { PersonSummary } from '../../store/persons/interfaces';
import PersonsList from './PersonsList';
import personsReducer from '../../store/persons/reducers';

const persons: PersonSummary[] = [
  {
    popularity: 20.346,
    known_for_department: 'Acting',
    gender: 2,
    id: 6384,
    profile_path: '/bOlYWhVuOiU6azC4Bw6zlXZ5QTC.jpg',
    adult: false,
    name: 'Keanu Reeves',
  },
  {
    popularity: 35.731,
    known_for_department: 'Directing',
    gender: 2,
    id: 224653,
    profile_path: '/b0fUF8XeZb3qDBb7ktEJlQL8ZAx.jpg',
    adult: false,
    name: 'Kazuo Kuroki',
  },
];

describe('PersonsList component', () => {
  let store: Store;

  it('renders correctly', () => {
    store = configureStore(rootReducer);

    const wrapper = mount(
      <Provider store={store}>
        <PersonsList />
      </Provider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should render Loading... when loading state is true', () => {
    store = configureStore(personsReducer, {
      persons: {
        page: 0,
        loading: true,
        error: false,
        persons: [],
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <PersonsList />
      </Provider>
    );

    const spanText = wrapper.find('span').text();

    expect(spanText).toEqual('Loading...');
  });

  it('should render RetryButton component when error state is true', () => {
    store = configureStore(personsReducer, {
      persons: {
        page: 0,
        loading: false,
        error: true,
        persons: [],
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <PersonsList />
      </Provider>
    );

    expect(wrapper.find('RetryButton')).toHaveLength(1);
  });

  it('should render persons list', () => {
    store = configureStore(personsReducer, {
      persons: {
        page: 0,
        loading: false,
        error: false,
        persons,
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <PersonsList />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('PersonShowcase')).toHaveLength(2);
  });
});
