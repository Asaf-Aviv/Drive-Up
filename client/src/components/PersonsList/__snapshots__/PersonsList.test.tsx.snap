// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonsList component renders correctly 1`] = `
"<Provider store={{...}}>
  <PersonsList>
    <main>
      <WithStyles(ForwardRef(List))>
        <ForwardRef(List) classes={{...}}>
          <ul className=\\"MuiList-root MuiList-padding\\">
            <span>
              Loading...
            </span>
          </ul>
        </ForwardRef(List)>
      </WithStyles(ForwardRef(List))>
    </main>
  </PersonsList>
</Provider>"
`;
